src/Type/ContainerDynamicReturnTypeExtension.php

```diff
@@ -41,24 +43,53 @@ public function getTypeFromMethodCall(
         Scope $scope
     ): Type {
         $returnType = ParametersAcceptorSelector::selectSingle($methodReflection->getVariants())->getReturnType();
-        $args = $methodCall->getArgs();
-        if (count($args) !== 1) {
-            return $returnType;
-        }
-
         $methodName = $methodReflection->getName();
-        $types = [];
-        $argType = $scope->getType($args[0]->value);
 
-        foreach ($argType->getConstantStrings() as $constantStringType) {
-            $serviceId = $constantStringType->getValue();
-            $service = $this->serviceMap->getService($serviceId);
-            if ($methodName === 'get') {
-                $types[] = $service !== null ? $service->getType() : $returnType;
-            } elseif ($methodName === 'has') {
+        if ($methodName === 'has') {
+            $args = $methodCall->getArgs();
+            if (count($args) !== 1) {
+                return $returnType;
+            }
+
+            $types = [];
+            $argType = $scope->getType($args[0]->value);
+
+            foreach ($argType->getConstantStrings() as $constantStringType) {
+                $serviceId = $constantStringType->getValue();
+                $service = $this->serviceMap->getService($serviceId);
                 $types[] = new ConstantBooleanType($service !== null);
             }
+
+            return TypeCombinator::union(...$types);
+        } elseif ($methodName === 'get') {
+            $args = $methodCall->getArgs();
+            if (count($args) === 0) {
+                return $returnType;
+            }
+
+            $types = [];
+
+            if (isset($args[1])) {
+                $invalidBehaviour = $scope->getType($args[1]->value);
+
+                if ($invalidBehaviour instanceof ConstantIntegerType
+                    && $invalidBehaviour->getValue() === ContainerInterface::NULL_ON_INVALID_REFERENCE
+                ) {
+                    $types[] = new NullType();
+                }
+            }
```

Comment:
To avoid an instance check, is there a matching getIntegerTypes or something to foreach over?

Reply:
found the `getConstantScalarValues` method on the `Type` interface.

This means it can handle compound types aswell now so great question